drivers/sensor/sgp40/sgp40.c

```diff
@@ -0,0 +1,309 @@
+/*
+ * Copyright (c) 2021 Leonard Pollak
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT sensirion_sgp40
+
+#include <device.h>
+#include <drivers/i2c.h>
+#include <kernel.h>
+#include <drivers/sensor.h>
+#include <sys/__assert.h>
+#include <logging/log.h>
+#include <sys/byteorder.h>
+#include <sys/crc.h>
+
+#include <drivers/sensor/sgp40.h>
+#include "sgp40.h"
+
+LOG_MODULE_REGISTER(SGP40, CONFIG_SENSOR_LOG_LEVEL);
+
+static uint8_t sgp40_compute_crc(uint16_t value)
+{
+	uint8_t buf[2];
+	sys_put_be16(value, buf);
+
+	return crc8(buf, 2, SGP40_CRC_POLY, SGP40_CRC_INIT, false);
+}
+
+static int sgp40_write_command(const struct device *dev, uint16_t cmd)
+{
+
+	const struct sgp40_config *cfg = dev->config;
+	uint8_t tx_buf[2];
+	sys_put_be16(cmd, tx_buf);
+
+	return i2c_write(cfg->bus, tx_buf, sizeof(tx_buf),
+			 cfg->i2c_addr);
+}
+
+static int sgp40_start_measurement(const struct device *dev)
+{
+	const struct sgp40_config *cfg = dev->config;
+	struct sgp40_data *data = dev->data;
+
+	uint8_t tx_buf[8] = { SGP40_CMD_MEASURE_RAW >> 8 ,
+		SGP40_CMD_MEASURE_RAW & 0xFF,
+		data->rh_param[0], data->rh_param[1], data->rh_param[2],
+		data->t_param[0], data->t_param[1], data->t_param[2],
+	};
+
+	return i2c_write(cfg->bus, tx_buf, sizeof(tx_buf),
+			 cfg->i2c_addr);
+}
+
+/*
+ * Set the params for RH/T compensation
+ */
+static int sgp40_update_compensation_params(const struct device *dev,
+			   int16_t t, int16_t rh)
+{
+	struct sgp40_data *data = dev->data;
+	int16_t rh_ticks, t_ticks;
+
+	sgp40_clamp(rh, SGP40_COMP_MIN_RH, SGP40_COMP_MAX_RH);
+	sgp40_clamp(t, SGP40_COMP_MIN_T, SGP40_COMP_MAX_T);
+
+	/*
+	 * Temperature and RH conversion to ticks as explained in datasheet
+	 * in section "I2C commands"
+	 */
+	rh_ticks = (rh * 0xFFFF / 100U) + 0.5;
+	t_ticks = ((t + 45U) * 0xFFFF / 175U) + 0.5;
+
+	data->rh_param[0] = rh_ticks >> 8;
+	data->rh_param[1] = rh_ticks & 0xFF;
+	data->rh_param[2] = sgp40_compute_crc(rh_ticks);
+
+	data->t_param[0] = t_ticks >> 8;
+	data->t_param[1] = t_ticks & 0xFF;
+	data->t_param[2] = sgp40_compute_crc(t_ticks);
+
+	return 0;
+}
+
+/*
+ * Public API for setting compensation data
+ */
```

Comment:
this function already has a docstring in header

Reply:
updated accordingly